pkg/api/unversioned.go

```diff
@@ -22,16 +22,64 @@ import (
 
 // This file contains API types that are unversioned.
 
-// APIVersions lists the api versions that are available, to allow
-// version negotiation. APIVersions isn't just an unnamed array of
-// strings in order to allow for future evolution, though unversioned
+// APIVersions lists the versions that are available, to allow clients to
+// discover the API at /api.
 type APIVersions struct {
+	// Versions are the api versions that are available.
 	Versions []string `json:"versions"`
 }
 
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+type APIGroupList struct {
+	// Groups is a list of APIGroup.
+	Groups []APIGroup `json:"groups"`
+}
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type APIGroup struct {
+	// GroupName represents the name of the group.
+	GroupName string `json:"groupName"`
+	// GroupVersions are the versions supported in this group.
+	GroupVersions []GroupVersion `json:"groupVersions"`
+	// PreferredVersion is the version preferred by the API server, which
+	// probably is the storage version.
+	PreferredVersion GroupVersion `json: "preferredVersion"`
+}
+
+// GroupVersion contains the "group/version" and "version" string of a version.
+// It is made a struct to keep extensiblity.
+type GroupVersion struct {
+	// GroupVersion is in the form of "group/version".
+	GroupVersion string `json:"groupVersion"`
+	// Version is in the form of "version", this is to save the clients the
+	// trouble of splitting the GroupVersion.
+	Version string `json:"version"`
+}
+
+// APIResource specifies the name of a resource and if it is namespaced.
+type APIResource struct {
+	// Resource is the name of the resource.
+	Resource string `json:"resource"`
```

Comment:
See the last bullet I recently added to the naming conventions:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions